### DISCUSSION TITLE
point cloud features

### DISCUSSION THREAD

**Discussion by AlexeySmolenchuk** [AcademySoftwareFoundation/openvdb#1370] [Q&A] [answered]
Hey guys,

I was pretty excited reading additional materials with performance tests of OpenVDB Points, so I decided to use it as point cloud format for rendering purposes, pretty much like Houdini Mantra does with .pc

I see there are some high-level structures for PointIndexGrid for example ParticleAtlas, which I assume might be useful. But I don't see any similar high-level API for PointDataGrid (which is the default output format for Houdini)

I want to implement a kind of search by distance, returning a pool of sorted points and corresponding attributes. Mainly random access. Ideally, from a file, without conversions.

Any suggestions and advice would be helpful.

Thanks,
Alexey

**Comment by Idclip** [comment, marked as answer]
Hey @somesanctus - there are currently no high level tools to achieve this for OpenVDB Points. The recently added PointRasterize tools do loop over neighbouring points based on search distances but are designed for transfering to numeric grids:

https://github.com/AcademySoftwareFoundation/openvdb/blob/2561c253d052aa27c5405d7fef4fcec1b2923ba0/openvdb/openvdb/points/PointTransfer.h#L460-L496

You could use something similar for a trivial search but may find that you need something more custom depending on your access patterns. Ultimately you currently need to implement this yourself - API for this is on the roadmap but no eta at the moment.

@danrbailey any thoughts?

**Reply by AlexeySmolenchuk** [reply]
Thank you

**Reply by danrbailey** [reply]
Good summary Nick, yes it has been on our list of things we wanted to write for a long time. The challenging aspect of this is that the points are already stored in a spatially organized way, but not in a way that's likely compatible with a desired neighbouring search so there's some investigation needed here. Perhaps a way of making progress here is simplify the scope of the problem and just ignore the existing spatial organisation and build a standard KD tree in order to establish the API then return to look at improvements to the algorithms and further optimizations.
